flowmachine/flowmachine/core/hour_slice.py

```diff
@@ -0,0 +1,258 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import re
+from operator import lt as less_than, ge as greater_or_equal
+from typing import List, Union
+
+from sqlalchemy.sql import func, and_, or_, true
+
+__all__ = ["HourInterval", "HourSlice"]
+
+
+class DayPeriod:
+    """
+    Represents a repeated "daily" time period.
+    """
+
+    def __init__(self, weekday=None):
+        self.freq = "day"
+        self.weekday = weekday
+        if self.weekday is not None:
+            raise ValueError(
+                "If freq='day' then the `weekday` argument must not be provided."
+            )
+
+    def filter_timestamp_column_by_day_of_week(self, ts_col):
+        # no additional filtering needed to limit the day of the week
+        return true()
+
+
+class DayOfWeekPeriod:
+    """
+    Represents a repeated "day-of-the-week" time period.
+    """
+
+    valid_weekdays = [
+        "Monday",
+        "Tuesday",
+        "Wednesday",
+        "Thursday",
+        "Friday",
+        "Saturday",
+        "Sunday",
+    ]
+    weekday_indices_postgres = {
+        "Sunday": 0,
+        "Monday": 1,
+        "Tuesday": 2,
+        "Wednesday": 3,
+        "Thursday": 4,
+        "Friday": 5,
+        "Saturday": 6,
+    }
+
+    def __init__(self, weekday):
+        if weekday is None:
+            raise ValueError(
+                "If freq='week' then the `weekday` argument must be provided."
+            )
+        weekday = weekday.capitalize()
+        if weekday not in self.valid_weekdays:
+            valid_weekdays_list = ", ".join([repr(x) for x in self.valid_weekdays])
+            raise ValueError(
+                f"Invalid value for `weekday`. Must be one of: {valid_weekdays_list}."
+            )
+
+        self.freq = "week"
+        self.weekday = weekday
+
+    def filter_timestamp_column_by_day_of_week(self, ts_col):
+        return (
+            func.extract("dow", ts_col) == self.weekday_indices_postgres[self.weekday]
+        )
+
+
+def make_hour_interval_period(freq, *, weekday=None):
+    """
+    Returns an appropriate instance of `DayPeriod` or `DayOfWeekPeriod`,
+    depending on the value of `freq`.
+    """
+    cls_lookup = {"day": DayPeriod, "week": DayOfWeekPeriod}
+
+    try:
+        cls = cls_lookup[freq]
+    except KeyError:
+        allowed_freqs = ", ".join([repr(x) for x in cls_lookup.keys()])
+        raise ValueError(f"Argument `freq` must be one of: {allowed_freqs}.")
+
+    return cls(weekday=weekday)
+
+
+class HourOfDay:
+    """
+    Represents an hour of the day and allows filtering a timestamp column accordingly.
+    """
+
+    def __init__(self, hour_str: Union[str, None]):
+        self._validate_hour_string(hour_str)
+        self.hour_str = hour_str
+
+    def __eq__(self, other):
```

Comment:
Why not just subclass `str` here?

Reply:
Frankly, I'm not entirely sure how this comparison would behave if we piggyback on `str` because `hour_str` can also be `None`, so I'd rather keep the explicit implementation here. But let me know if you have a strong preference for subclassing and know how to tweak the `__eq__` method to make it work.

Reply:
Is the None check required anywhere aside from the `filter_timestamp_column` method?

Reply:
Nope. And now that I think of it that should perhaps actually live in a separate sister class representing a `MissingHourAndMinutesTimestamp` or similar...

Admittedly, the `__eq__` comparison is also mainly a convenience for the tests (albeit a nice one to have). So we _could_ get rid of it if you prefer.

Reply:
Well that's the nice thing about subclassing string - you stick the validation in `__new__` and you have a string that is guaranteed to have that particular formatting and the one additional method on it. If that's the _only_ place the comparison against `None` is used, you can do the compare against `'None'`.

Reply:
Well, I'm comparing against None in the tests, so that needs to happen in the `__eq__` method. But maybe I should just get rid of that and do an explicit comparison with the full object...

Reply:
Actually, your suggestion to subclass from string works nicely! 👍

I have introduced a new class `MissingHourAndMinutesTimestamp` which is returned from `__new__` if the input value is `None`, which separates the responsibilities better and is actually reasonably elegant and readable. :) Let me know if you're happy with this approach.